libs/cpp-ledger/ledger/ledger_device_hid.cpp

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "cpp-ledger/ledger/ledger_device_hid.hpp"
+
+#include "cpp-ledger/ledger/apdu_wrapper.hpp"
+#include "cpp-ledger/ledger/const.hpp"
+#include "cpp-ledger/ledger/utils.hpp"
+
+namespace ledger {
+
+  std::tuple<Bytes, Error> LedgerDeviceHid::Exchange(
+      const Bytes &command) const {
+    if (command.size() < 5) {
+      return std::make_tuple(
+          Bytes{}, Error{"APDU commands should not be smaller than 5"});
+    }
+
+    if (command.size() - 5 != command[4]) {
+      return std::make_tuple(Bytes{}, Error{"APDU[data length] mismatch"});
+    }
+
+    auto [serializedCommand, err] =
+        apdu::WrapCommandApdu(kChannel, command, kPacketSize);
+    if (err != std::nullopt) {
+      return std::make_tuple(Bytes{}, err);
+    }
+
+    // Write all the packets
+    std::tie(std::ignore, err) = Write(serializedCommand);
+    if (err != std::nullopt) {
+      return std::make_tuple(Bytes{}, err);
+    }
+
+    const auto readData = Read();
+
+    const auto [response, err1] =
+        apdu::UnwrapResponseApdu(kChannel, readData, kPacketSize);
+    if (err1 != std::nullopt) {
+      return std::make_tuple(Bytes{}, err1);
+    }
+
+    if (response.size() < 2) {
+      return std::make_tuple(Bytes{}, Error{"response length < 2"});
+    }
+
+    const auto swOffset = response.size() - 2;
```

Comment:
What stands for sw?

Reply:
I don't know. It is name of return code from origin library.
Basically, it is the name of ADPU return codes from specification: https://www.eftlab.co.uk/knowledge-base/complete-list-of-apdu-responses/